### DISCUSSION TITLE
OMEdit command-line interface checks?

### DISCUSSION THREAD

**Discussion by bilderbuchi** [OpenModelica/OpenModelica#10113] [Q&A] [answered]
I'm not sure if that is worth an issue, but I'm wondering if OMEdit has a formal command-line interface? If yes, is it true and desired that arguments are not checked?

I've been using it to start the new WIP API with `--NAPI=true`. Then I realised that I was not able to find out what switches are available (trying `--help`, `-h`), and more importantly, that OMEdit does not reject invalid arguments, and just starts up.
For example, ` C:\OpenModelica1.21.0-dev-64bit\bin\OMEdit.exe --hahalol` would just start OMEdit instead of erroring about an invalid argument. 
I found that very curious.

**Comment by casella** [comment]
You can set default and model-specific compilation (a.k.a. translation) and simulation flags, using Tools | Options | Simulation and the Simulation setup menus.

You can also interact with omc in the background via a commadn-line interface Tools | OpenModelica Compiler CLI, using the scripting language and the [OMC API](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/scripting_api.html). However, this can have unexpected side effects when combined with the way OMEdit internally interacts with the compiler in the background, so handle with care.

The scripting language is not really formally defined anywhere, and mostly used for testing. If you want to do serious scripting, you can use OMPython.

**Reply by bilderbuchi** [reply]
Ah, I was not sure if I could add that inside OMEdit, I thought it had to be started like you suggested to do.

This was purely about starting OMEdit with the new api, no scripting intended.

**Comment by perost** [comment]
I'm not sure if it's different on Windows somehow, but at least on Linux the `--help` flag does work. Although it only lists two flags: `--Debug` and `--NAPI`. Mostly I think we try to avoid command line flags for OMEdit since it's a graphical program.

But I have also noticed that OMEdit will still start even if you use an incorrect flag, but at least it prints an error to the terminal in that case. Maybe we should log such error in OMEdit's message browser too to make it more visible, or possibly even refuse to start entirely. Maybe @adeas31 has some opinions on that.

**Reply by bilderbuchi** [reply]
> I'm not sure if it's different on Windows somehow, but at least on Linux the --help flag does work. 

I just checked again on Windows, it does not.

> But I have also noticed that OMEdit will still start even if you use an incorrect flag, but at least it prints an error to the terminal in that case.

On Windows, nothing is printed.

Yeah, I would suggest refusing to start as a sane behaviour if given invalid arguments.

**Comment by adeas31** [comment, marked as answer]
As @perost mentioned since OMEdit is a GUI app so we don't care too much about the command line options. The command line options are mostly for developers. The `-NAPI` option will be removed or moved to settings once we make this default.

`--help` should print the information but windows command prompt will only show the message if we make OMEdit a console application i.e., `CONFIG += console` (https://doc.qt.io/qt-6/qmake-variable-reference.html#config). I don't want to do this since this will always pop up a windows console even if OMEdit is launched from start menu or by double clicking the executable. There are hacks and workarounds available to fix these things but there are also known issues with them.

HINT: Use a better terminal like MSYS instead of Windows command prompt to see the output.

**Reply by bilderbuchi** [reply]
Thanks, I understand. I agree, it's not worth it to enable console mode for this, considering the drawbacks. Non-console mode doesn't stop you from quitting when encountering invalid flags though, or does it?

I guess I was just surprised to find quite nonstandard behaviour again. 

> HINT: Use a better terminal like MSYS instead of Windows command prompt to see the output.

Nice dig at poor Windows users ;-)
Just FYI, I'm using PowerShell, not command prompt on Windows, does that qualify as "better"? (Ubuntu at home).

**Reply by adeas31** [reply]
> Thanks, I understand. I agree, it's not worth it to enable console mode for this, considering the drawbacks. Non-console mode doesn't stop you from quitting when encountering invalid flags though, or does it?

I guess quitting incase of invalid flags will give us the same problem that Windows users won't see any output. Maybe its better to print a warning in OMEdit's `Messages Browser` about invalid flag and continue running the GUI as normal.

> 
> I guess I was just surprised to find quite nonstandard behaviour again.
> 
> > HINT: Use a better terminal like MSYS instead of Windows command prompt to see the output.
> 
> Nice dig at poor Windows users ;-) Just FYI, I'm using PowerShell, not command prompt on Windows, does that qualify as "better"? (Ubuntu at home).

PowerShell is more strange. It neither print the output nor does it redirect the output. Atleast with command prompt you can run `OMEdit.exe --help > a.txt`

**Reply by bilderbuchi** [reply]
😆 🤷

**Reply by adeas31** [reply]
#10124 will report the use of invalid flags in the Messages Browser. Other than that there is nothing more I can do.

**Reply by casella** [reply]
LGTM
